### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by LeonIllya**

### CODE REVIEW

File: src/main/java/book/store/dto/cartitem/CartItemDto.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package book.store.dto.cartitem;
+
+import lombok.Data;
+
+@Data
+public class CartItemDto {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/cartitem/CartItemDto.java:6]: You should have CartItemRequestDto (bookId, quantity), UpdateCartItemDto(quantity) and response CartItemDto(id, bookId, bookTitle, quantity).

File: src/main/java/book/store/mapper/CartItemMapper.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+import book.store.config.MapperConfig;
+import book.store.dto.cartitem.CartItemDto;
+import book.store.dto.cartitem.CreateCartItemRequestDto;
+import book.store.model.CartItem;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
+public interface CartItemMapper {
+    CartItemDto toDto(CartItem cartItem);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/mapper/CartItemMapper.java:11]:
```suggestion
    @Mapping(source = "book.id", target = "bookId")
    @Mapping(source = "book.title", target = "bookTitle")
    CartItemDto toDto(CartItem cartItem);
```

File: src/main/java/book/store/mapper/ShoppingCartMapper.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package book.store.mapper;
+
+import book.store.config.MapperConfig;
+import book.store.dto.shoppingcart.ShoppingCartDto;
+import book.store.model.ShoppingCart;
+import org.mapstruct.Mapper;
+
+@Mapper(config = MapperConfig.class)
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/mapper/ShoppingCartMapper.java:8]:
```suggestion
@Mapper(config = MapperConfig.class, uses = CartItemMapper.class)
```

File: src/main/java/book/store/model/CartItem.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Getter
+@Setter
+@EqualsAndHashCode(of = {"id"})
+@ToString
+@Entity(name = "cart_items")
+@SQLDelete(sql = "UPDATE cart_items SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/CartItem.java:23]: `@Where` is deprecated, use [SQLRestriction](https://docs.jboss.org/hibernate/stable/orm/javadocs/org/hibernate/annotations/Where.html) instead

File: src/main/java/book/store/model/CartItem.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/CartItem.java:28]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/book/store/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/CartItem.java:31]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/book/store/model/CartItem.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    @JoinColumn(name = "book_id",nullable = false)
+    private Book book;
+    @Column(nullable = false)
+    private int quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/CartItem.java:35]:
```suggestion
    private int quantity;
    private boolean isDeleted; ? 
```

File: src/main/java/book/store/model/ShoppingCart.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+@Where(clause = "is_deleted=false")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
+    private Set<CartItem> cartItems;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/ShoppingCart.java:33]:
```suggestion
    private Set<CartItem> cartItems;
    private boolean isDeleted; ?
```

File: src/main/java/book/store/model/ShoppingCart.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+@EqualsAndHashCode(of = {"id"})
+@ToString
+@Entity(name = "shopping_carts")
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/ShoppingCart.java:29]:
```suggestion
    @OneToOne(fetch = FetchType.LAZY)
```

File: src/main/java/book/store/model/ShoppingCart.java [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart")
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/ShoppingCart.java:32]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

File: src/main/java/book/store/model/ShoppingCart.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+@Entity(name = "shopping_carts")
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/ShoppingCart.java:31]: You can use `@MapsId`

File: src/main/java/book/store/repository/shoppingcart/ShoppingCartRepository.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+import book.store.model.ShoppingCart;
+import book.store.validation.Email;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query(value = "SELECT sc.user FROM ShoppingCart sc JOIN sc.user u WHERE u.email = :email",
+            nativeQuery = true)
+    Optional<ShoppingCart> findUserByEmail(@Email String email);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/repository/shoppingcart/ShoppingCartRepository.java:12]:
```suggestion
   ShoppingCart findByUserId(Long userId);
```

File: src/main/java/book/store/service/impl/CartItemServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+import book.store.repository.book.BookRepository;
+import book.store.repository.cartitem.CartItemRepository;
+import book.store.service.CartItemService;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@RequiredArgsConstructor
+@Service
+public class CartItemServiceImpl implements CartItemService {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/CartItemServiceImpl.java:18]: please, change logic according to comment in ShoppingCartController.

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's improve your solution!

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+        if (cartItem.getShoppingCart().getUser().getEmail().equals(email)) {
+            cartItem.setQuantity(quantity);
+            cartItemRepository.save(cartItem);
+        }
+    }
+
+    @Override
+    public void deleteBookById(Long cartId, String userEmail) {
+        CartItem cartItem = cartItemRepository.findById(cartId).orElseThrow(() ->
+            new EntityNotFoundException("Can`t find cart item by id: " + cartId));
```
Comment (okuzan) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:47]: extract repeated logic to a private method

File: src/main/java/book/store/repository/shoppingcart/ShoppingCartRepository.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+
+import book.store.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query(value = "SELECT sc FROM ShoppingCart sc JOIN sc.user u WHERE u.id = :userId",
+            nativeQuery = true)
+    ShoppingCart findByUserId(Long userId);
+
```
Comment (okuzan) [inline src/main/java/book/store/repository/shoppingcart/ShoppingCartRepository.java:11]:
```suggestion
```

File: src/main/java/book/store/repository/shoppingcart/ShoppingCartRepository.java [hunk 16]
```diff
@@ -0,0 +1,12 @@
+package book.store.repository.shoppingcart;
+
+import book.store.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query(value = "SELECT sc FROM ShoppingCart sc JOIN sc.user u WHERE u.id = :userId",
+            nativeQuery = true)
```
Comment (okuzan) [inline src/main/java/book/store/repository/shoppingcart/ShoppingCartRepository.java:9]: try to write this in JPQL, not native query

File: src/main/java/book/store/model/ShoppingCart.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@SQLRestriction(value = "is_deleted = false")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", nullable = false)
+    @MapsId
+    private User user;
```
Comment (okuzan) [inline src/main/java/book/store/model/ShoppingCart.java:36]: do you need all of them? seems like you create user_id column which will have the same value as id of shopping cart, seems redundant

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Shopping cart management", description = "Endpoint for managing shopping cart")
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/cart")
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java:31]:
```suggestion
@RequestMapping("/cart")
```

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+        User user = (User) authentication.getPrincipal();
+        return shoppingCartService.addBookInShoppingCart(user.getId(), cartItemRequestDto);
+    }
+
+    @PutMapping("/cart-items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    public void updateQuantity(Authentication authentication,
+                                            @PathVariable @Positive Long cartItemId,
+                                            @RequestBody @Valid UpdateCartItemDto item) {
+
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java:56]:
```suggestion
```

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Shopping cart management", description = "Endpoint for managing shopping cart")
+@RequiredArgsConstructor
+@RestController
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java:30]:
```suggestion
@RestController
@Validated
```

File: src/main/java/book/store/repository/shoppingcart/ShoppingCartRepository.java [hunk 21]
```diff
@@ -0,0 +1,11 @@
+package book.store.repository.shoppingcart;
+
+import book.store.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query("SELECT sc FROM ShoppingCart sc WHERE sc.user.id = :id")
+    ShoppingCart getUserById(@Param("id") Long userId);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/repository/shoppingcart/ShoppingCartRepository.java:10]: Why your method called getUserById if you select shopping cart by user id? 
findByUserId - is better

File: src/main/java/book/store/dto/cartitem/CartItemRequestDto.java [hunk 22]
```diff
@@ -0,0 +1,9 @@
+package book.store.dto.cartitem;
+
+import lombok.Data;
+
+@Data
+public class CartItemRequestDto {
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/cartitem/CartItemRequestDto.java:6]: Add validation

File: src/main/java/book/store/dto/cartitem/UpdateCartItemDto.java [hunk 23]
```diff
@@ -0,0 +1,8 @@
+package book.store.dto.cartitem;
+
+import lombok.Data;
+
+@Data
+public class UpdateCartItemDto {
+    private int quantity;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/dto/cartitem/UpdateCartItemDto.java:7]: Add validation

File: src/main/java/book/store/service/ShoppingCartService.java [hunk 24]
```diff
@@ -0,0 +1,15 @@
+package book.store.service;
+
+import book.store.dto.cartitem.CartItemRequestDto;
+import book.store.dto.shoppingcart.ShoppingCartDto;
+import jakarta.validation.constraints.Positive;
+
+public interface ShoppingCartService {
+    ShoppingCartDto addBookInShoppingCart(Long userId, CartItemRequestDto requestDto);
+
+    void updateQuantity(String cartId, @Positive Long cartItemId, int quantity);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/ShoppingCartService.java:10]:
```suggestion
    ShoppingCartDto updateQuantity(String cartId,  Long cartItemId, UpdateCartItemDto .....);
```
It's better to validate in controller, dto and entity

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+        CartItem cartItem = findCartItemById(cartItemId);
+
+        if (cartItem.getShoppingCart().getUser().getEmail().equals(userEmail)) {
+            cartItemRepository.deleteById(cartItem.getId());
+        }
+    }
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Long userId) {
+        ShoppingCart byUserId = shoppingCartRepository.getUserById(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:54]: byUserId bad naming

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:24]:
```suggestion
    @Override
    @Transactional
```

Check other places where is neede

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+    public ShoppingCartDto addBookInShoppingCart(Long userId, CartItemRequestDto requestDto) {
+        ShoppingCart userShoppingCart = shoppingCartRepository.getUserById(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:26]: Do we need to check if book exists in DB? Do we need to set book in cartItem? 
What if the book is already in shopping cart, maybe we should increase it quantity? 

Please, check several cases by adding new book and see the response, adding the same book also.

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 28]
```diff
@@ -0,0 +1,62 @@
...
+        CartItem model = cartItemMapper.toModel(requestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:27]: model is bad naming

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+        cartItemRepository.save(model);
+        userShoppingCart.getCartItems().add(model);
+        return shoppingCartMapper.toDto(userShoppingCart);
+    }
+
+    @Override
+    public void updateQuantity(String email, Long cartItemId, int quantity) {
+        CartItem cartItem = findCartItemById(cartItemId);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:35]: Don't need to use email, you have PK id to find what you need. 
```
ShoppingCart cart = repository.findByUserId(userId);
CartItem cartItem = itemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())...
```

File: src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+
+        if (cartItem.getShoppingCart().getUser().getEmail().equals(email)) {
+            cartItem.setQuantity(quantity);
+            cartItemRepository.save(cartItem);
+        }
+    }
+
+    @Override
+    public void deleteBookById(Long cartItemId, String userEmail) {
+        CartItem cartItem = findCartItemById(cartItemId);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/ShoppingCartServiceImpl.java:45]: You can use the same as above comment
```
ShoppingCart shoppingCart = repository.findByUserId(userId);
CartItem cartItem = itemRepository.findByIdAndShoppingCartId(itemId, shoppingCart.getId())
```

File: src/main/java/book/store/controller/ShoppingCartController.java [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+    @PreAuthorize("hasRole('USER')")
+    public void deleteBookById(Authentication authentication,
+                                @PathVariable @Positive Long cartItemId) {
+        User user = (User) authentication.getPrincipal();
+        shoppingCartService.deleteBookById(cartItemId, user.getEmail());
+    }
+
+    @GetMapping("/cart-items/{cartItemId}")
+    @PreAuthorize("hasRole('USER')")
+    public CartItemDto getCartItemByCartId(Authentication authentication,
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/ShoppingCartController.java:72]: Why do you need this endpoint?

File: src/main/resources/db/changelog/changes/10-create-shopping-cart-table.yaml [hunk 32]
```diff
@@ -0,0 +1,28 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-shopping-cart-table
+      author: LeonIlya
+      changes:
+        - createTable:
+            tableName: shopping_carts
+            columns:
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/10-create-shopping-cart-table.yaml:8]:
```suggestion
            columns:
             - column:
                  name: id
                  type: bigint
                  constraints:
                    primaryKey: true
                    nullable: false
              - column:
                  name: is_deleted
                  type: boolean
                  defaultValueBoolean: false
                  constraints:
                    nullable: false
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's improve your solution!

**Review by sarakhmen — APPROVED** [review]
Good job!

**Pull request merged**